processing/src/main/java/org/apache/druid/segment/incremental/OakIncrementalIndex.java

```diff
@@ -0,0 +1,684 @@
+    int dimType = buff.getInt(dimIndexInBuffer);
+    if (dimType == OakIncrementalIndex.NO_DIM) {
+      return null;
+    } else if (dimType == ValueType.DOUBLE.ordinal()) {
+      dimObject = buff.getDouble(dimIndexInBuffer + OakIncrementalIndex.DATA_OFFSET);
+    } else if (dimType == ValueType.FLOAT.ordinal()) {
+      dimObject = buff.getFloat(dimIndexInBuffer + OakIncrementalIndex.DATA_OFFSET);
+    } else if (dimType == ValueType.LONG.ordinal()) {
+      dimObject = buff.getLong(dimIndexInBuffer + OakIncrementalIndex.DATA_OFFSET);
+    } else if (dimType == ValueType.STRING.ordinal()) {
+      int arrayIndexOffset = buff.getInt(dimIndexInBuffer + OakIncrementalIndex.ARRAY_INDEX_OFFSET);
+      int arrayIndex = buff.position() + arrayIndexOffset;
+      int arraySize = buff.getInt(dimIndexInBuffer + OakIncrementalIndex.ARRAY_LENGTH_OFFSET);
+      int[] array = new int[arraySize];
+      for (int i = 0; i < arraySize; i++) {
+        array[i] = buff.getInt(arrayIndex);
+        arrayIndex += Integer.BYTES;
+      }
+      dimObject = array;
+    }
+
+    return dimObject;
+  }
+
+
+  static class AggsManager
+  {
+    private final AggregatorFactory[] metrics;
+    private final ReentrantLock[] aggLocks;
+    private volatile Map<String, ColumnSelectorFactory> selectors;
+    private final boolean reportParseExceptions;
+
+    //given a ByteBuffer and an offset where all aggregates for a row are stored
+    //offset + aggOffsetInBuffer[i] would give position in ByteBuffer where ith aggregate
+    //is stored
+    private volatile int[] aggOffsetInBuffer;
+    private volatile int aggsTotalSize;
+    private final BufferAggregator[] aggs;
+
+    public AggsManager(BufferAggregator[] aggs,
+                       Map<String, ColumnSelectorFactory> selectors,
+                       int[] aggOffsetInBuffer,
+                       int aggsTotalSize, AggregatorFactory[] metrics,
+                       boolean reportParseExceptions)
+    {
+      this.aggs = aggs;
+      this.selectors = selectors;
+      this.aggOffsetInBuffer = aggOffsetInBuffer;
+      this.aggsTotalSize = aggsTotalSize;
+      this.metrics = metrics;
+      this.aggLocks = new ReentrantLock[this.aggs.length];
+      for (int i = 0; i < this.aggLocks.length; i++) {
+        this.aggLocks[i] = new ReentrantLock(true);
+      }
+      this.reportParseExceptions = reportParseExceptions;
+    }
+
+    public void initValue(ByteBuffer byteBuffer,
+                          InputRow row,
+                          ThreadLocal<InputRow> rowContainer)
+    {
+      if (metrics.length > 0 && aggs[0] == null) {
+        // note: creation of Aggregators is done lazily when at least one row from input is available
+        // so that FilteredAggregators could be initialized correctly.
+        rowContainer.set(row);
+        for (int i = 0; i < metrics.length; i++) {
+          final AggregatorFactory agg = metrics[i];
+          aggLocks[i].lock();
```

Comment:
remove locks ?

Reply:
Cannot remove locks. the init of aggs is lazy and only done once. just like multithreaded singleton creation